core/src/main/scala/akka/kafka/internal/ProducerStage.scala

```diff
@@ -121,32 +121,79 @@ private[kafka] object ProducerStage {
       }
     })
 
-    def produce(msg: Message[K, V, P]): Unit = {
-      val r = Promise[Result[K, V, P]]
-      awaitingConfirmation.incrementAndGet()
-      producer.send(msg.record, new Callback {
-        override def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
-          if (exception == null) {
-            onMessageAckCb.invoke(msg)
-            r.success(Result(metadata, msg))
-          }
-          else {
-            decider(exception) match {
-              case Supervision.Stop =>
-                if (stage.closeProducerOnStop) {
-                  producer.close(0, TimeUnit.MILLISECONDS)
+    def produce(in: Messages[K, V, P]): Unit = {
+      in match {
+        case msg: Message[K, V, P] =>
+          val r = Promise[Result[K, V, P]]
+          awaitingConfirmation.incrementAndGet()
+          producer.send(msg.record, new Callback {
+            override def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
+              if (exception == null) {
+                onMessageAckCb.invoke(msg)
+                r.success(Result(metadata, msg))
+              }
+              else {
+                decider(exception) match {
+                  case Supervision.Stop =>
+                    if (stage.closeProducerOnStop) {
+                      producer.close(0, TimeUnit.MILLISECONDS)
+                    }
+                    failStageCb.invoke(exception)
+                  case _ =>
+                    r.failure(exception)
                 }
-                failStageCb.invoke(exception)
-              case _ =>
-                r.failure(exception)
+              }
+
+              if (awaitingConfirmation.decrementAndGet() == 0 && inIsClosed)
+                checkForCompletionCB.invoke(())
             }
+          })
+          val future = r.future.asInstanceOf[Future[OUT]]
+          push(stage.out, future)
+
+        case multiMsg: MultiMessage[K, V, P] =>
```

Comment:
It seems like there's an opportunity here to generalize these two cases to DRY things up.  A `Messages` that is not a `PassThroughMessage` represents 1..m messages to produce.  Maybe `Messages` could have an internal method to get a collection of messages which can be mapped to a produce method.  It could also contain a factory method to generate a `Results`.

Reply:
Yes, but I think the different mechanics around firing off many produce calls an gathering the results are so much more complicated than the single message ones, that it is OK to have it kind of duplicated.

Reply:
I agree that DRYing it up completely would bring in additional not needed complexity. However the `Callback` to the `producer.send` is almost completely (1 line difference) identical. That could be factored out.

Reply:
Yes, I think `producer.send` is the best opportunity too.  I'm definitely in favour in keeping things simple and readable.